test-runner-frontend/src/Views/Dashboard.js

```diff
@@ -5,377 +5,146 @@ import Stack from '@mui/material/Stack';
-      improvedVsRegressed: {
-        improved: 8,
-        regressed: 7,
-      },
-      machine: machinesArray[Math.floor(Math.random() * machinesArray.length)],
-      date: '12/08/2022 8:59 pm',
-    },
-    {
-      id: 'a1224f86-1981-11ed-861d-0242ac120002',
-      branch: branchArray[Math.floor(Math.random() * branchArray.length)],
-      status: statusArray[Math.floor(Math.random() * statusArray.length)],
-      benchmarks: 10,
-      improvedVsRegressed: {
-        improved: 5,
-        regressed: 5,
-      },
-      machine: machinesArray[Math.floor(Math.random() * machinesArray.length)],
-      date: '12/08/2022 9:12 pm',
-    },
-    {
-      id: 'af2ac950-1981-11ed-861d-0242ac120002',
-      branch: branchArray[Math.floor(Math.random() * branchArray.length)],
-      status: statusArray[Math.floor(Math.random() * statusArray.length)],
-      benchmarks: 7,
-      improvedVsRegressed: {
-        improved: 4,
-        regressed: 3,
-      },
-      machine: machinesArray[Math.floor(Math.random() * machinesArray.length)],
-      date: '12/08/2022 10:42 pm',
-    },
-  ]);
+  useEffect(
+    () => () =>
+      onSnapshot(collection(db, FIRESTORE_COLLECTION_NAME), (snapshot) =>
+        dispatch({ type: 'testResults', payload: snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id })) })
+      ),
+    []
+  );
 
-  /*
-  const buildRows = () => {
-    testResults?.sort((a, b) => b.testRunnerResults.core_32.startTime - a.testRunnerResults.core_32.startTime);
-    testResults?.map((test, index) => {
-      var timestamp = test.testRunnerResults.core_32.startTime;
-      var date = new Date(timestamp * 26);
-      const newObj = {
-        id: test.commitHash,
-        //branch: textArray[Math.floor(Math.random()*textArray.length)],
-        status: test.status.core_32,
-        benchmarks: test.testRunnerResults.core_32.testStatistics?.length,
-        improvedVsRegresed: compareBenchmarks(testResults[index], testResults[index + 1] ? testResults[index + 1] : []),
-        machine: machinesArray[Math.floor(Math.random() * machinesArray.length)],
-        date: date.toLocaleString(),
-      }
-      setTestRunResults(testRunResults => [...testRunResults, newObj]);
-      setTestRunResultsCopy(testRunResultsCopy => [...testRunResultsCopy, newObj]);
-      console.log(newObj.improvedVsRegresed);
-    })
+  useEffect(() => {
+    compareBenchmarks(testResults);
+    // eslint-disable-next-line
+  }, [testResults]);
+
+  const compareBenchmarks = (testRunResults) => {
+    const machines = getMachineNames(testRunResults);
+    const sortedTestRuns = sortTestRunsByMachine(machines, testRunResults);
+    const benchmarksByMachine = getBenchmarksByMachine(sortedTestRuns);
+    const allBenchmarks = getAllBenchmarks(benchmarksByMachine);
+    dispatch({ type: 'testRunResultsCopy', payload: allBenchmarks });
+    dispatch({ type: 'testRunResults', payload: allBenchmarks });
+    dispatch({ type: 'loadingState', payload: allBenchmarks });
```

Comment:
part of the joy of using a reducer is that we don't need to do three independent `dispatch()` calls. we can decide on a type (maybe compare) and send `allBenchmarks` as the payload. then within the reducer, we can handle the compare type and set `testRunResultsCopy`, `testRunResults` and `loadingState`. let me know if you want to pair on the reducers

Reply:
I think I was able to implement it, but if my solutions is not what you meant, I would really appreciate your help! 👍